### PR TITLE
feat(kit): new `DateTime` mask

### LINKED ISSUE #86: 🚀 - `DateTime` -mask
### Which package(s) are relevant/related to the feature request?

@maskito/kit

### Description

We should create mask for [InputDateTime](https://taiga-ui.dev/components/input-date-time).

In the most cases the new mask should behave in the same way as `Date` + `Time` masks.

## `@Input()`-s

* **dateMode**
* **timeMode**
* **dateSeparator**
* **min**
(btw, native `Date` can accepts hours, minutes ...)
* **max**

### PR DESCRIPTION
**Pull request by vladimirpotekhin**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #86 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/demo-integrations/cypress/tests/kit/date-time/date-time-basic.cy.ts [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+                    .should('have.prop', 'selectionStart', '10.0'.length)
+                    .should('have.prop', 'selectionEnd', '10.0'.length);
+            });
+
+            it('11.11.2011, 1|2:3|0 => Delete => 11.11.2011, 10:0|0', () => {
+                cy.get('@input')
+                    .type('111120111230')
+                    .should('have.value', '11.11.2011, 12:30')
+                    .type('{leftArrow}'.repeat('0'.length))
+                    .realPress(['Shift', ...Array('1.1'.length).fill('ArrowLeft')]);
```
Comment (nsbarsukov) [inline projects/demo-integrations/cypress/tests/kit/date-time/date-time-basic.cy.ts:263]:
```suggestion
                    .realPress(['Shift', ...Array('2:3'.length).fill('ArrowLeft')]);
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    header="DateTime"
+    package="KIT"
+>
+    <ng-template pageTab>
+        Use
+        <code>maskitoDateTimeOptionsGenerator</code>
+        to create mask to input a date and time.
+
+        <tui-doc-example
+            id="Localization"
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html:11]:
```suggestion
            id="localization"
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            [content]="dateTimeLocalization"
+            [description]="localizationDescription"
+        >
+            <ng-template #localizationDescription>
+                Use
+                <code>dateMode</code>
+                ,
+                <code>timeMode</code>
+                and
+                <code>separator</code>
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html:22]:
```suggestion
                <code>dateSeparator</code>
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                parameters to get mask with a language‑sensitive representation
+                of dates.
+            </ng-template>
+            <date-time-mask-doc-example-1></date-time-mask-doc-example-1>
+        </tui-doc-example>
+
+        <tui-doc-example
+            id="MinMax"
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html:30]:
```suggestion
            id="min-max"
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        <tui-doc-demo [control]="apiPageControl">
+            <ng-template>
+                <tui-input
+                    tuiTextfieldCustomContent="tuiIconCalendarLarge"
+                    [tuiTextfieldFiller]="
+                        getPlaceholder(dateMode, timeMode, dateSeparator)
+                    "
+                    [formControl]="apiPageControl"
+                >
+                    Enter dates
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html:65]:
```suggestion
                    Enter date and time
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+                documentationPropertyType="MaskitoTimeMode"
+                [documentationPropertyValues]="timeModeOptions"
+                [(documentationPropertyValue)]="timeMode"
+                (documentationPropertyValueChange)="updateOptions()"
+            >
+                Time format mode
+            </ng-template>
+
+            <ng-template
+                documentationPropertyName="separator"
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.template.html:97]:
```suggestion
                documentationPropertyName="dateSeparator"
```

File: projects/demo/src/pages/kit/date-time/date-time-mask-doc.module.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import {DateTimeMaskDocExample2} from './examples/2-min-max/component';
+
+@NgModule({
+    imports: [
+        CommonModule,
+        FormsModule,
+        ReactiveFormsModule,
+        MaskitoModule,
+        TuiAddonDocModule,
+        TuiHintModule,
```
Comment (nsbarsukov) [inline projects/demo/src/pages/kit/date-time/date-time-mask-doc.module.ts:21]: Should we drop this module ?

File: projects/kit/src/lib/masks/date-time/utils/parse-date-time-string.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    dateTime: string,
+    dateModeTemplate: string,
+): string[] {
+    const hasSeparator = dateTime.includes(', ');
+
+    return [
+        dateTime.slice(0, dateModeTemplate.length),
+        dateTime.slice(
+            hasSeparator
+                ? dateModeTemplate.length + ', '.length
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/date-time/utils/parse-date-time-string.ts:11]:
```suggestion
                ? dateModeTemplate.length + DATE_TIME_SEPARATOR.length
```

File: projects/kit/src/lib/masks/date-time/utils/parse-date-time-string.ts [hunk 9]
```diff
@@ -0,0 +1,15 @@
+export function parseDateTimeString(
+    dateTime: string,
+    dateModeTemplate: string,
+): string[] {
+    const hasSeparator = dateTime.includes(', ');
```
Comment (nsbarsukov) [inline projects/kit/src/lib/masks/date-time/utils/parse-date-time-string.ts:5]:
```suggestion
    const hasSeparator = dateTime.includes(DATE_TIME_SEPARATOR);
```

File: projects/kit/src/lib/utils/time/parse-time-string.ts [hunk 10]
```diff
@@ -1,9 +1,9 @@
-import {getObjectFromEntries} from '../../../utils';
-import {MaskitoTimeSegments} from '../types';
-
 /**
  * @param timeString can be with/without fixed characters
  */
+import {MaskitoTimeSegments} from '../../types';
+import {getObjectFromEntries} from '../get-object-from-entries';
```
Comment (nsbarsukov) [inline projects/kit/src/lib/utils/time/parse-time-string.ts:5]: revert this file ?

File: projects/demo-integrations/cypress/tests/kit/date-time/date-time-separator.cy.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
+import {DemoPath} from '@demo/path';
+
+describe('DateTime | Separator', () => {
+    describe('/', () => {
+        beforeEach(() => {
+            cy.visit(`/${DemoPath.DateTime}/API?separator=/`);
```
Comment (nsbarsukov) [inline projects/demo-integrations/cypress/tests/kit/date-time/date-time-separator.cy.ts:6]:
```suggestion
            cy.visit(`/${DemoPath.DateTime}/API?dateSeparator=/`);
```

Comment (nsbarsukov) [reply projects/demo-integrations/cypress/tests/kit/date-time/date-time-separator.cy.ts]: @vladimirpotekhin there are other `separator => dateSeparator` cases in `*.cy.ts`-files (

File: projects/demo-integrations/cypress/tests/kit/date-time/date-time-separator.cy.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+                .type('1412.')
+                .should('have.value', '14/12')
+                .type('2000')
+                .should('have.value', '14/12/2000');
+        });
+    });
+
+    describe('-', () => {
+        beforeEach(() => {
+            cy.visit(`/${DemoPath.DateRange}/API?separator=-`);
```
Comment (nsbarsukov) [inline projects/demo-integrations/cypress/tests/kit/date-time/date-time-separator.cy.ts:29]:
```suggestion
            cy.visit(`/${DemoPath.DateRange}/API?dateSeparator=-`);
```

**Review by nsbarsukov — APPROVED** [review]

File: projects/kit/src/lib/utils/date/date-to-segments.ts [hunk 13]
```diff
@@ -1,9 +1,13 @@
...
-export function dateToSegments(date: Date): MaskitoDateSegments {
+export function dateToSegments(date: Date): MaskitoDateSegments & MaskitoTimeSegments {
     return {
         day: String(date.getDate()).padStart(2, '0'),
         month: String(date.getMonth() + 1).padStart(2, '0'),
         year: String(date.getFullYear()).padStart(4, '0'),
+        hours: String(date.getHours()).padStart(2, '0'),
+        minutes: String(date.getMinutes()).padStart(2, '0'),
+        seconds: String(date.getSeconds()).padStart(2, '0'),
+        milliseconds: String(date.getMilliseconds()).padStart(2, '0'),
```
Comment (waterplea) [inline projects/kit/src/lib/utils/date/date-to-segments.ts:11]: Shouldn't it be `padStart(3, '0')`?

File: projects/kit/src/lib/masks/date-time/postprocessors/min-max-date-time-postprocessor.ts [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+}: {
+    dateModeTemplate: string;
+    timeMode: MaskitoTimeMode;
+    min?: Date;
+    max?: Date;
+}): NonNullable<MaskitoOptions['postprocessor']> {
+    return ({value, selection}) => {
+        const [dateString, timeString] = parseDateTimeString(value, dateModeTemplate);
+
+        if (!isDateTimeStringCompleted(value, dateModeTemplate, timeMode)) {
```
Comment (waterplea) [inline projects/kit/src/lib/masks/date-time/postprocessors/min-max-date-time-postprocessor.ts:29]: `Complete` without `d` is probably more grammatically correct.

**Review by waterplea — APPROVED** [review]

File: projects/kit/src/lib/utils/date/segments-to-date.ts [hunk 15]
```diff
@@ -1,11 +1,18 @@
-import {MaskitoDateSegments} from '../../types';
+import {MaskitoDateSegments, MaskitoTimeSegments} from '../../types';
 
-export function segmentsToDate(parsedDate: Partial<MaskitoDateSegments>): Date {
+export function segmentsToDate(
+    parsedDate: Partial<MaskitoDateSegments>,
+    parsedTime?: Partial<MaskitoTimeSegments>,
+): Date {
     const year = parsedDate.year?.length === 2 ? `20${parsedDate.year}` : parsedDate.year;
```
[the remark this answers is not in the document]
Comment (vladimirpotekhin) [reply projects/kit/src/lib/utils/date/segments-to-date.ts:7]: This line is for detecting short/long format of the year 

but you are right, we can't write first 99 years for another reason 🤔

[the remark this answers is not in the document]
Comment (vladimirpotekhin) [reply projects/kit/src/lib/utils/date/segments-to-date.ts:7]: fixed

**Review by nsbarsukov — APPROVED** [review]

**Pull request merged**